addons/mail/models/mail_mail.py

```diff
@@ -81,6 +89,42 @@ def default_get(self, fields):
     scheduled_date = fields.Char('Scheduled Send Date',
         help="If set, the queue manager will send the email after the date. If not set, the email will be send as soon as possible.")
 
+    def _compute_message_int(self):
+        for mail in self:
+            mail.message_int = mail.mail_message_id.id
+
+    @api.depends('attachment_ids')
+    def _compute_restricted_attachments(self):
+        """We might not have access to all the attachments of the emails.
+        Compute the attachments we have access to,
+        and the number of attachments we do not have access to.
+        """
+        for mail_sudo, mail in zip(self.sudo(), self):
+            mail.unrestricted_attachment_ids = mail._filter_attachment_access()
+            mail.restricted_attachment_count = len(mail_sudo.attachment_ids) - len(mail.unrestricted_attachment_ids)
+
+    def _inverse_unrestricted_attachment_ids(self):
+        """We can only remove the attachments we have access to."""
+        for mail_sudo, mail in zip(self.sudo(), self):
+            restricted_attaments = mail_sudo.attachment_ids - mail._filter_attachment_access()
+            mail_sudo.attachment_ids = restricted_attaments | mail.unrestricted_attachment_ids
+
+    def _filter_attachment_access(self):
+        """Filter the given attachment to return only the records the current user have access to."""
+        ret_attachments = self.env['ir.attachment']
+        if not self.check_access_rights('read', raise_exception=False):
+            return ret_attachments
+
+        for attachment in self.sudo().attachment_ids:
+            # Use SUDO here to not raise an error during the prefetch
+            # And then drop SUDO right to check if we can access it
+            try:
+                attachment.sudo(False).check('read')
```

Comment:
Shame that ``check`` does not support working like ``_filter_access_rules`` who keep only valid records ... I suppose you have to do a per-record basis because it raises.

Do you think this would be much work to have it available on attachment model (would mean somehow re-implementing the check / extracting common parts, ...). Asking for a friend :)

Reply:
it's a bit tricky because you read read acl to execute the method on the records (and if we execute in SUDO, we should see all attachment, etc)

but if it can be useful, what I can do is a model method, call it with the attachments ids in arguments

what do you think ?